primitives/session/src/offchain_hashdb.rs

```diff
@@ -0,0 +1,312 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Database storage based on the offchain index db for historical
+//! slashing behaviour.
+//!
+//! For being able to erase per session data, there exist per
+//! `session_index` keyd index
+//! which tracks all entries which are associated with the `session_index`.
+//!
+//! The mapping is as follows
+//! ```markdown
+//! session -> [(derived_key,tree_prefix),(derived_key,tree_prefix),..]
+//! ```
+//! where `derived_key` is a unique identifier derived from the `digest`
+//! of the associated data stored, `tree_prefix` as well as `session_index`.
+//! This allows direct re-use for pruning individual values.
+//!
+//! The implementation contains two (2) different key deriving functions,
+//! one for the index and one for the actual key value pairs in order
+//! to disambiguate values in the underlying key value storage layer.
+//!
+//! Encoding and decoding is achieved by [`codec`](self::codec::Codec).
+
+use trie_db::TrieLayout;
+
+use hash_db::{AsHashDB, HashDB, Hasher, Prefix};
+
+use codec::{Decode, Encode};
+use sp_core::offchain::StorageKind;
+use sp_io::offchain;
+use sp_std::prelude::*;
+use sp_std::collections::btree_set::BTreeSet;
+
+#[derive(Default)]
+pub struct AlternativeDB<L>
+where
+    L: TrieLayout + Send,
+{
+    session_index: Vec<u8>,
+    _phantom: core::marker::PhantomData<L>,
+}
+
+// @todo requires some internal mutability concept to work properly
+// @todo which is yet to be hashed out
+unsafe impl<L> core::marker::Sync for AlternativeDB<L> where
```

Comment:
nothing in the struct definition makes me think this unsafe implementation is needed. Generally we avoid unsafe code very strongly, so I'd rather take a performance hit than introduce unsafe code.

Reply:
Unfortuantely the trait bound of `HashDB` requires it to be `sync`.

Reply:
I don't see how a `Sync` bound, in the absence of some other factors, implies interior mutability. I haven't looked at this in some time, but don't the functions of `HashDB` that require you to mutate pass `&mut self`? If that is the case, then are you saying that there is a reason we would need interior mutability on the `&self` methods of `HashDB`?